api/src/main/java/org/openmrs/layout/name/NameSupport.java

```diff
@@ -27,16 +48,143 @@ public NameSupport() {
 	}
 	
 	public static NameSupport getInstance() {
-		if (singleton == null) {
-			throw new APIException("Not Yet Instantiated");
-		} else {
-			return singleton;
+		synchronized (NameSupport.class) {
+			singleton.init();
+			if (singleton.customTemplateConfigured) {
+				return singleton;
+			} else {
+				new NameSupport();
+			}
+			
+			if (singleton == null) {
+				throw new APIException("Not Yet Instantiated");
+			} else {
+				return singleton;
+			}
+		}
+	}
+
+	private void init() {
+		if (!initialized) {
+			Context.getAdministrationService().addGlobalPropertyListener(singleton);
+			try {
+				// Get layout NameTemplate format to override the default if any
+				String layoutTemplateXml = Context.getAdministrationService().getGlobalProperty(
+					OpenmrsConstants.GLOBAL_PROPERTY_LAYOUT_NAME_TEMPLATE);
+				
+				if (StringUtils.isNotBlank(layoutTemplateXml)) {
+					// Validating... Custom NameTemplate should be in a Valid XML format
+					DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
+					DocumentBuilder builder = dbf.newDocumentBuilder();
+					Document doc = builder.parse(new ByteArrayInputStream(layoutTemplateXml.getBytes(StandardCharsets.UTF_8)));
+					
+					setNameTemplate(layoutTemplateXml);
+					List<String> specialTokens = new ArrayList<>();
+					specialTokens.add("prefix");
+					specialTokens.add("givenName");
+					specialTokens.add("middleName");
+					specialTokens.add("familyNamePrefix");
+					specialTokens.add("familyName");
+					specialTokens.add("familyName2");
+					specialTokens.add("familyNameSuffix");
+					specialTokens.add("degree");
+					setSpecialTokens(specialTokens);
+					customTemplateConfigured = true;
+					
+				}
+			} catch (ParserConfigurationException | IOException | SAXException e) {
+				log.error("Custom NameTemplate should be in a Valid XML format", e);
+			} catch (NullPointerException ignored) {}
+			initialized = true;
 		}
 	}
+
+	/**
+	 * @param nameTemplates The nameTemplates to set.
+	 */
+	public void setNameTemplate(List<NameTemplate> nameTemplates) {
+		this.layoutTemplates = nameTemplates;
+	}
```

Comment:
I'd rather have this changed to `addLayoutTemplate(NameTemplate nameTemplate) {...}`. Let's add the custom template to the existing templates and not completely reset them.